File: services/appointment-service/src/models/Appointment.js
// appointment-service/src/models/appointment.js
const mongoose = require('mongoose');

const AppointmentSchema = new mongoose.Schema({
  patientName: {
    type: String,
    required: [true, 'Patient name is required'],
    trim: true,
    index: true
  },
  patientContact: {
    email: {
      type: String,
      trim: true,
      lowercase: true,
      match: [/^\S+@\S+\.\S+$/, 'Please enter a valid email address']
    },
    phone: {
      type: String,
      trim: true
    }
  },
  doctorId: {
    type: String, // Store as string to avoid MongoDB ObjectId dependency
    required: [true, 'Doctor ID is required'],
    index: true
  },
  doctorName: {
    type: String,
    required: [true, 'Doctor name is required'],
    trim: true
  },
  date: {
    type: Date,
    required: [true, 'Appointment date is required'],
    index: true
  },
  time: {
    type: String,
    required: [true, 'Appointment time is required'],
    match: [/^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/, 'Time format should be HH:MM']
  },
  type: {
    type: String,
    required: [true, 'Appointment type is required'],
    enum: {
      values: ['Consultation', 'Follow-up', 'Emergency', 'Test', 'Procedure'],
      message: '{VALUE} is not a valid appointment type'
    }
  },
  status: {
    type: String,
    enum: {
      values: ['Scheduled', 'Checked-in', 'In-progress', 'Completed', 'Cancelled', 'No-show'],
      message: '{VALUE} is not a valid status'
    },
    default: 'Scheduled',
    index: true
  },
  notes: {
    type: String,
    trim: true
  },
  duration: {
    type: Number,
    min: 5,
    default: 30, // duration in minutes
    required: true
  },
  cancellationReason: {
    type: String,
    trim: true
  },
  cancellationDate: {
    type: Date
  },
  reminderSent: {
    type: Boolean,
    default: false
  },
  // Track who created and last updated the appointment
  createdBy: {
    type: String,
    default: 'system'
  },
  updatedBy: {
    type: String,
    default: 'system'
  }
}, { 
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Add virtual for formatting date in YYYY-MM-DD format
AppointmentSchema.virtual('formattedDate').get(function() {
  return this.date ? this.date.toISOString().split('T')[0] : null;
});

// Text search index
AppointmentSchema.index({ patientName: 'text' });

// Compound indexes for common queries
AppointmentSchema.index({ doctorId: 1, date: 1 });
AppointmentSchema.index({ status: 1, date: 1 });
AppointmentSchema.index({ date: 1, time: 1 });

module.exports = mongoose.model('Appointment', AppointmentSchema);

File: frontend/lib/features/appointment/models/appointment.dart
import 'package:flutter/foundation.dart';

class Appointment {
  final String? id;
  final String patientName;
  final String? patientEmail;
  final String? patientPhone;
  final String doctorId;
  final String doctorName;
  final DateTime date;
  final String time;
  final String type;
  final String status;
  final String? notes;
  final int duration;
  final String? cancellationReason;
  final bool reminderSent;

  Appointment({
    this.id,
    required this.patientName,
    this.patientEmail,
    this.patientPhone,
    required this.doctorId,
    required this.doctorName,
    required this.date,
    required this.time,
    required this.type,
    this.status = 'Scheduled',
    this.notes,
    this.duration = 30,
    this.cancellationReason,
    this.reminderSent = false,
  });

  factory Appointment.fromJson(Map<String, dynamic> json) {
    try {
      DateTime parsedDate;
      if (json['date'] is String) {
        parsedDate = DateTime.parse(json['date']);
      } else if (json['date'] is Map) {
        parsedDate = DateTime.parse(json['date']['\$date'] ?? json['date']['date']);
      } else {
        parsedDate = DateTime.now();
      }

      final patientContact = json['patientContact'] as Map<String, dynamic>?;

      return Appointment(
        id: json['_id'] ?? json['id'],
        patientName: json['patientName'] ?? '',
        patientEmail: patientContact?['email'],
        patientPhone: patientContact?['phone'],
        doctorId: json['doctorId'] ?? '',
        doctorName: json['doctorName'] ?? '',
        date: parsedDate,
        time: json['time'] ?? '00:00',
        type: json['type'] ?? 'Consultation',
        status: json['status'] ?? 'Scheduled',
        notes: json['notes'],
        duration: json['duration'] is int
            ? json['duration']
            : int.tryParse('${json['duration']}') ?? 30,
        cancellationReason: json['cancellationReason'],
        reminderSent: json['reminderSent'] == true,
      );
    } catch (e) {
      debugPrint('Error parsing appointment JSON: $e');
      debugPrint('Raw JSON: $json');
      rethrow;
    }
  }

  Map<String, dynamic> toJson() {
    final contact = <String, dynamic>{};
    if (patientEmail != null && patientEmail!.isNotEmpty) {
      contact['email'] = patientEmail;
    }
    if (patientPhone != null && patientPhone!.isNotEmpty) {
      contact['phone'] = patientPhone;
    }

    return {
      if (id != null) 'id': id,
      'patientName': patientName,
      if (contact.isNotEmpty) 'patientContact': contact,
      'doctorId': doctorId,
      'doctorName': doctorName,
      'date': date.toIso8601String(),
      'time': time,
      'type': type,
      'status': status,
      'notes': notes,
      'duration': duration,
      'cancellationReason': cancellationReason,
      'reminderSent': reminderSent,
    };
  }
}
